Change gc for nim's fast implementation to recommended arc gc

OpenSourceAnarchist opened the pull request:
While markAndSweep is mostly better than the default gc in nim, the new "arc" gc is even more performant and results in a lot less memory usage overall. In order to be consistent with new upstream development, the "fast" nim implementation should use arc. More details available at https://nim-lang.org/docs/gc.html

See https://youtu.be/yA32Wxl59wo?t=830 for an example benchmark. Using arc instead of markAndSweep saves over 10 seconds and uses less memory!

OpenSourceAnarchist committed:
Change gc for main_fast to recommended arc gc

OpenSourceAnarchist committed:
Update README with changes to gc settings

frol commented:
@OpenSourceAnarchist Thank you for the contribution! I will get back to it once I have enough free time. Sorry for the delay, the things are hectic at the moment.

OpenSourceAnarchist committed:
Update README.md

Disabling runtime safety checks tends to offer noticeable speed improvements if your code is written safely

OpenSourceAnarchist committed:
Update Makefile

Disabling runtime safety checks tends to offer noticeable speed improvements if your code is written safely

OpenSourceAnarchist commented:
These other changes disable runtime safety checks for the "tuned" Nim implementations. In reality this is obviously not a great idea for most code, but since this is a benchmark, it's worth trying out :)

If the Nim code doesn't run or compile, I'd try reverting the gc (arc) commits before the `-d:danger` ones, but both should not break any code. It is recommended to run the Nim nightly/devel version for using the arc gc mode, so if you're still on Arch it might benefit you to use the nim-git AUR package (https://aur.archlinux.org/packages/nim-git/) instead of the 1.2.0 version in the community repo, but it shouldn't make that much of a difference.

Fixes #89

frol commented:
I am not going to include stats for dev/nightly branches of compilers, as I don't have the bandwidth to deal with them, sorry. You can provide your measurements for the benchmarked versions on the solutions from the scoreboard alongside the dev/nightly results on your hardware as a comment in an issue as a reference for those who are interested in those comparisons and I will add a link to the README.

OpenSourceAnarchist committed:
Don't use ARC gc yet

OpenSourceAnarchist committed:
Don't use ARC gc yet

OpenSourceAnarchist commented:
I benchmarked a variety of configurations using the latest nightly version on a Windows 10 desktop. Using the ARC gc option significantly reduced performance, so I got rid of those changes in this PR. However, the `-d:danger` flag did slightly improve the runtime performance.

The results (no memory stats, just seconds from before calling the main function to directly after, averages of 100 runs for each config option):

naive: 0.4369715857505798s
naive + markAndSweep: 0.2204265379905701s
fast + markAndSweep: 0.1612434196472168s
manual + none: 0.1611993479728699s


naive + danger + markAndSweep: 0.1983912968635559s
naive + arc: 0.36674724817276s
naive + danger + arc: 0.4587049174308777s
fast + danger + markAndSweep: 0.159391508102417s
fast + danger + arc: 0.3393148064613342s

Hardware: Intel i5-4690k @ 3.50 GHz, 16 GB ram, Windows 10 Home version 1909, OS build 18363.778, Nim version 1.3.5 (devel)

OpenSourceAnarchist commented:
Closing as I am no longer keeping up with Nim development. If this project is ever updated, I'll make another updated PR for Nim with new benchmarks, and submit an implementation in V (similar goals to Nim, but with Go-like syntax and compile-time memory management).
